Xamarin.Forms.Platform.iOS/CollectionView/CarouselViewController.cs

```diff
@@ -202,13 +202,40 @@ void CollectionItemsSourceChanged(object sender, NotifyCollectionChangedEventArg
 
 			_gotoPosition = -1;
 
-			if (count > 0)
-				ScrollToPosition(newPosition, carouselPosition, false);
+			bool removingCurrentItem = e.Action == NotifyCollectionChangedAction.Remove &&
+									currentItemPosition == -1;
+
+			ScrollToPositionIfNeeded(newPosition, count, removingCurrentItem);
 
 			SetCurrentItem(newPosition);
 			SetPosition(newPosition);
 		}
 
+		void ScrollToPositionIfNeeded(int newPosition, int count, bool removingCurrentItem)
+		{
+			// if we don't have items no need to scroll
+			if (count <= 0)
+				return;
+
+			if (!removingCurrentItem)
+			{
+				ScrollToPosition(newPosition, Carousel.Position, false);
+			}
+			else
+			{
+				// we will Reloadthe VisibleItems so we need delay a bit for the collectionView to reload
+				// and then scroll to the correct position
+				Device.StartTimer(TimeSpan.FromMilliseconds(50), () =>
```

Comment:
Don't we have any other callback mechanism we can use to determine the right time to do this? 50ms seems arbitrary. What happens if the reload takes longer?

Reply:
Yeah I know, on my research I didn't found one, there are other hacks around to get try to know this.. The reload is just for visible items not for the entire collection.

Reply:
Ok @hartez  think i got it to work with PerformBatchUpdates.  so first time i tried i did something like 
```
CollectionView.PerformBatchUpdates(() => CollectionView.ReloadItems(IndexPathsForsVisibleItems), (_) => ScrollToPosition(newPosition, Carousel.Position, false, removingCurrentElement));
```
But that didn't work,  if i passed something in the update Action. So today i tested again and if put the reload outside of PerformBatchUpdates and then call PerformBatchUpdates and use the callback it seems to work, maybe it sets the Scroll in the correct order on the Main thread ? 

```
 CollectionView.ReloadItems(IndexPathsForsVisibleItems);
CollectionView.PerformBatchUpdates(() => {}, (_) => ScrollToPosition(newPosition, Carousel.Position, false, removingCurrentElement));
```